### DISCUSSION TITLE
Ensure Gramine SGX app is called only by another specific Gramine SGX app avoiding two extra ra-tls-servers

### DISCUSSION THREAD

**Discussion by tiagorvmartins** [gramineproject/gramine#1619] [Q&A] [answered]
I have been looking into the example at [ra-tls-mbedtls](https://github.com/gramineproject/gramine/tree/master/CI-Examples/ra-tls-mbedtls), which seems to be a good starting point for what I am trying to achieve, however some questions below.

Assuming I have two gramine sgx apps that talk to each other, I would need two extra RA-TLS-Servers running on SGX that would confirm that each sgx enclave gramine app running have the expected MRENCLAVE, is this correct?

Continuing on this, what if I want to prevent access to an gramine enclave sgx (that exposes an endpoint), such that only a specific sgx process with a specific mrenclave can do it?

Assuming I want to avoid having the two extra RA-TLS-Servers - I have Gramine SGX App A and Gramine SGX App B, A calls B, and I need to ensure that only enclave A with MRENCLAVE_A is the one and only able to reach and communicate with Gramine SGX App B. 

Could I feed in the RA_TLS_env_vars to Sgx App B such that would verify and validate that only MRENCLAVE_A is able to call it? if this was possible basically my Sgx App B would be working as a ra-tls-server aswell, I believe?

Also by the way, I can't really change underlying code/app that exposes an endpoint on the Sgx App B,
Not sure if this makes sense at all, still learning the gist of it, thank you in advance!

**Comment by mkow** [comment]
On the conceptual level, this is achievable using the low-level attestation interface of Gramine: https://gramine.readthedocs.io/en/latest/attestation.html#low-level-dev-attestation-interface. You can manually locally-attest the other enclave, exchange encryption keys and set up a TLS channel bound to these keys.

> A calls B, and I need to ensure that only enclave A with MRENCLAVE_A is the one and only able to reach and communicate with Gramine SGX App B.

> I can't really change underlying code/app that exposes an endpoint on the Sgx App B

Hmm, am I missing something? You say that B is exposing an unprotected endpoint and you can't change it, but you actually want to change it to not accept any connections other than from A? Isn't this contradictory?
Remember that Gramine can't alter the way your app is implementing TLS endpoints, we can't modify your TLS logic on-the-fly and bind the encryption keys to the attestation (which is the only way to ensure you're talking to a specific enclave).

**Reply by tiagorvmartins** [reply]
Thank you for the reply @mkow, my lack of knowledge didn't help explaining it properly. 

Enclave B is based on a binary and won't be adapted, it actually exposes an http endpoint by default, I thought that I could use gramine sgx in a way thay would wrap the underlying connection such that it could only be made by another specific client gramine sgx enclave, but that's not the goal of gramine, sorry for my mistake. 

I understand now thats something out of scope and different subject. 

To make it clear: in case of any code changes on client enclave A (different mrenclave), that client shouldn't be able to reach enclave b. My end goal is preventing that access. Based on the client mrenclave. 

So if I understood correctly, I believe this involves two steps.

1) a reverse proxy that would proxy https connections to the enclave B http endpoint
2) implement the low level attestation on the reverse proxy enclave in a way that would verify that any incoming tls connection is from client enclave A

This is because, like you said, gramine need to work on top of an app that serves tls endpoints and that can be changed/configurable. Using step 1, I would have a new enclave (the reverse proxy) in between the two enclaves (A and B) 

Does it make sense? 
Do you believe this could work this way?
Thank you again.

**Reply by kailun-qin** [reply]
> a reverse proxy that would proxy https connections to the enclave B http endpoint

@tiagorvmartins This sounds like the enclave B http endpoint can already be restricted from reaching/accessing at the network level? Otherwise, w/ the above proposal, how can the enclave B http endpoint prevent from being directly accessed (i.e., not thru the reverse proxy)?

**Reply by dimakuv** [reply]
> a reverse proxy that would proxy https connections to the enclave B http endpoint

I understand your problem, @tiagorvmartins, but this reverse-proxy design is broken. The problem is that your reverse-proxy will still forward the data to Enclave B in plaintext (over plain HTTP). I guess that's not what you want.

Generally, Gramine is not expected to work with non-TLS applications. In the sense that if your application establishes non-TLS connections (e.g. this HTTP endpoint), then Gramine can't help here.

Basically, you must ensure that the TLS endpoint terminates **inside the same enclave that will process the data**. I think there are some libraries that could help with that, but Gramine doesn't provide any "connection wrapper/encrypter" out of the box.

---

That said, if you're Ok forwarding plaintext data to your enclave B, and your Reverse Proxy enclave is only used to whitelist "good enclaves" (like Enclave A with known MRENCLAVE), and your enclave B is **not accessible** from the network but only from the Reverse Proxy, then your solution could work.

**Reply by dimakuv** [reply]
@tiagorvmartins You may read this short paper about the TLS-endpoint-termination problem in SGX: https://www.doc.ic.ac.uk/research/technicalreports/2017/DTRS17-5.pdf

I think the proposed library in that paper (TaLoS) was only a Proof of Concept and not stable and not really usable, but the paper gives a good overview of the problem.

**Reply by dimakuv** [reply]
@tiagorvmartins Finally, a tool and an example were added to Gramine recently (will be available in the next release):
- https://gramine.readthedocs.io/en/latest/manpages/gramine-ratls.html
- https://github.com/gramineproject/gramine/tree/master/CI-Examples/ra-tls-nginx

This gives you ideas how to apply RA-TLS to an application in a transparent manner. However, the application (in this case the Nginx web server) **must be TLS-enabled**.

**Reply by tiagorvmartins** [reply]
Thank you all for the detailed replies and discussions,
I will try to answer each message.

@kailun-qin yes that's correct, enclave b running the wrapped binary in gramine can be launched via a specific flag that will restrict access to only a specific address, and this could the address of the reverse proxy application (nginx) - such that the only http traffic incoming to enclave b would be the one from the reverse proxy.

@dimakuv yes I understand the reverse-proxy design problem being broken, but I guess I won't have any other choice here. Reverse Proxy enclave and Enclave B can be placed on the same machine itself to minimize http plain data leaks over the web, also reverse proxy  will the only able to send http plain data to the enclave b by using the flag mentioned above.

@dimakuv thank you for sharing the paper on the TLS-endpoint-termination! will take a look!

Also thank you @dimakuv for the example on gramine-ratls, haven't notice this one before.
I guess my last piece of learning is understanding how to whitelist on nginx only incoming https connection from enclave A, and I believe this might be using ra-tls as you mentioned, following what is proposed on using the environment variables (check copy snippet below) [ra-tls-mbedtls ](https://github.com/gramineproject/gramine/tree/master/CI-Examples/ra-tls-mbedtls#ra-tls-client), might be the way forward for my use case, I believe?

`If client is run without additional command-line arguments, it uses default RA-TLS verification callback that compares MRENCLAVE, MRSIGNER, ISV_PROD_ID and ISV_SVN against the corresonding RA_TLS_* environment variables. MRENCLAVE, MRSIGNER and ISV_PROD_ID are expected to match RA_TLS_* ones.`

Thank you again for all the input everyone.

**Reply by dimakuv** [reply, marked as answer]
> I guess my last piece of learning is understanding how to whitelist on nginx only incoming https connection from enclave A, and I believe this might be using ra-tls as you mentioned, following what is proposed on using the environment variables (check copy snippet below) [ra-tls-mbedtls ](https://github.com/gramineproject/gramine/tree/master/CI-Examples/ra-tls-mbedtls#ra-tls-client), might be the way forward for my use case, I believe?

Yes, you are correct.
